core/src/Kokkos_Half.hpp

```diff
@@ -1328,7 +1346,227 @@ struct Kokkos::Experimental::Impl::max_exponent_helper<
   static constexpr int value = 128;
 };
 #endif
-////////////// END BHALF_T (bfloat16) limits //////////////
+////////////// END BHALF_T (bfloat16) limits //////////
+
+#if defined(KOKKOS_ENABLE_SYCL)
+#define KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE sycl
+#else
+#if (defined(KOKKOS_COMPILER_NVCC) || defined(KOKKOS_COMPILER_NVHPC)) && \
+    defined(__GNUC__) && (__GNUC__ < 6) && !defined(__clang__)
+#define KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE
+#else
+#define KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE std
+#endif
+#endif
+
+namespace Kokkos {
+////////////// BEGIN HALF_T (float16) MATH FNS ////////////
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC)                \
+  KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t FUNC(       \
+      Kokkos::Experimental::half_t x) {                           \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;             \
+    return static_cast<Kokkos::Experimental::half_t>(             \
+        FUNC(static_cast<float>(x)));                             \
+  }                                                               \
+  KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t FUNC##fp16( \
```

Comment:
We probably shouldn't provide these: there is no clear convention for this (for example NVIDIA has hexp not exp_fp16, and uses hexp for both half and bhalf. So lets leave this overload out.